### DISCUSSION TITLE
`RuntimeError` with custom middleware when client drops connection before request finished

### DISCUSSION THREAD

**Discussion by santilade** [encode/starlette#1527] [Potential Issue] [answered]
## The issue
When using two or more custom middleware that subclass `BaseHTTPMiddleware`, if the client drops the connection during a request, `RuntimeError("No response returned.")` is raised.

## How to reproduce
Create the following file 
```python
# example.py:
import asyncio

from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
from starlette.routing import Route


async def homepage(request):
    await asyncio.sleep(5)
    return JSONResponse({"hello": "world"})


routes = [
    Route("/", homepage, name="homepage"),
]


class CustomHeaderMiddleware1(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        return await call_next(request)


class CustomHeaderMiddleware2(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        return await call_next(request)


all_middleware = [
    Middleware(CustomHeaderMiddleware1),
    Middleware(CustomHeaderMiddleware2),
]


app = Starlette(debug=True, routes=routes, middleware=all_middleware)
```
Run server:
```
uvicorn example:app  
```
Send HTTP request and disconnect before response is received:
```
curl http://127.0.0.1:8000 -m 2
```

You then get the following error on the server: 
```
INFO:     Started server process [70987]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/site-packages/anyio/streams/memory.py", line 81, in receive
    return self.receive_nowait()
  File "/site-packages/anyio/streams/memory.py", line 76, in receive_nowait
    raise WouldBlock
anyio.WouldBlock

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/site-packages/starlette/middleware/base.py", line 41, in call_next
    message = await recv_stream.receive()
  File "/site-packages/anyio/streams/memory.py", line 101, in receive
    raise EndOfStream
anyio.EndOfStream

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/site-packages/uvicorn/protocols/http/httptools_impl.py", line 378, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc
  File "/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/site-packages/starlette/middleware/base.py", line 63, in __call__
    response = await self.dispatch_func(request, call_next)
  File "./starlette_test.py", line 22, in dispatch
    return await call_next(request)
  File "/site-packages/starlette/middleware/base.py", line 45, in call_next
    raise RuntimeError("No response returned.")
RuntimeError: No response returned.

```

Is there something we are doing wrong or is this an issue?

**Comment by illagrenan** [comment]
I have noticed exactly the same problem with `starlette==0.18.0` and `uvicorn==0.17.5`. I found out that one of my clients started disconnecting prematurely and the server log was completely overwhelmed by `RuntimeError`.

I am currently using following middleware as a quickfix (this middleware needs to be registered first). The middleware basically just catches the `RuntimeError` and checks that the client is disconnected - and then returns a dummy response (which the client doesn't read anyway).

```python
# -*- encoding: utf-8 -*-
# ! python3

from __future__ import annotations

import logging
from http import HTTPStatus
from typing import Final

from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response

logger: Final = logging.getLogger(__name__)


class SuppressNoResponseReturnedMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        try:
            response = await call_next(request)
        except RuntimeError as e:
            if await request.is_disconnected() and str(e) == "No response returned.":
                logger.warning(
                    "Error `No response returned` detected. "
                    "At the same time we know that the client is disconnected "
                    "and not waiting for a response."
                )

                return Response(status_code=HTTPStatus.NO_CONTENT)
            else:
                raise

        return response
```

**Comment by kprzybyla** [comment]
Me and my team stumbled upon the same issue couple of days ago and implemented the same kind of middleware as a workaround (it has to be the last middleware that was installed by the call `add_middleware`). We spent too much time trying to figure out why `RuntimeError("No response returned.")` is raised thinking that we did something wrong in our middlewares or the external middlewares were buggy but after the investigation it turned out to be a client timeout issue.

I even got a pull request ready with the correction and was about to start discussion on this but you beat me to it.

Anyway, I think that the main issue here is that the error is misleading and below correction in the `startlette` code should be sufficient to make it clear what is happening and the users can catch `ClientDisconnect` error in their middlewares if custom behaviour/reaction to the client disconnect is needed.

```diff
diff --git a/starlette/middleware/base.py b/starlette/middleware/base.py
index bfb4a54..4da49aa 100644
--- a/starlette/middleware/base.py
+++ b/starlette/middleware/base.py
@@ -2,7 +2,7 @@ import typing
 
 import anyio
 
-from starlette.requests import Request
+from starlette.requests import Request, ClientDisconnect
 from starlette.responses import Response, StreamingResponse
 from starlette.types import ASGIApp, Receive, Scope, Send
 
@@ -42,6 +42,8 @@ class BaseHTTPMiddleware:
             except anyio.EndOfStream:
                 if app_exc is not None:
                     raise app_exc
+                if await request.is_disconnected():
+                    raise ClientDisconnect("Client disconnect while processing the request")
                 raise RuntimeError("No response returned.")
 
             assert message["type"] == "http.response.start"
```

**Comment by guillaume-chervet** [comment]
We have the same issue with 

```
fastapi==0.76.0
uvicorn==0.17.6
```

The workaround seems to work.

**Reply by felipecwb** [reply]
didn't worked for me...

steps to replicate:
1. create a long request like 10s
2. do a cURL and cut/close the request in the middle
3. should show error logs

**Comment by felipecwb** [comment]
I got the same here:

Requirements:
```
anyio==3.5.0
starlette==0.19.1
fastapi==0.77.1
uvicorn[standard]==0.17.6
```
Exception:
```
Traceback (most recent call last):
  File "uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi
              result = await app(self.scope, self.receive, self.send)
  File "uvicorn/middleware/proxy_headers.py", line 75, in __call__
          return await self.app(scope, receive, send)
  File "fastapi/applications.py", line 269, in __call__
          await super().__call__(scope, receive, send)
  File "starlette/applications.py", line 124, in __call__
          await self.middleware_stack(scope, receive, send)
  File "starlette/middleware/errors.py", line 184, in __call__
              raise exc
  File "starlette/middleware/errors.py", line 162, in __call__
              await self.app(scope, receive, _send)
  File "MY_APPLICATION/api/middlewares/prometheus/asgi.py", line 100, in __call__
              return await self.app(scope, receive, send)
  File "starlette/middleware/cors.py", line 84, in __call__
              await self.app(scope, receive, send)
  File "starlette/middleware/base.py", line 68, in __call__
              response = await self.dispatch_func(request, call_next)
  File "MY_APPLICATION/api/middlewares/etag.py", line 31, in middleware
          return await call_next(request)
  File "starlette/middleware/base.py", line 47, in call_next
                  raise RuntimeError("No response returned.")
RuntimeError: No response returned.

Traceback (most recent call last):
  File "starlette/middleware/base.py", line 43, in call_next
                  message = await recv_stream.receive()
  File "anyio/streams/memory.py", line 101, in receive
                  raise EndOfStream
anyio.EndOfStream

Traceback (most recent call last):
  File "anyio/streams/memory.py", line 81, in receive
              return self.receive_nowait()
  File "anyio/streams/memory.py", line 76, in receive_nowait
          raise WouldBlock
anyio.WouldBlock
```

**Comment by bob-rove** [comment]
Quite a few people complaining about exactly this problem in different issue reports around Starlette, FastAPI and other frameworks using the underlying toolset.

See for example:
- https://github.com/encode/starlette/issues/1255

For many peeps it's really hard to relate what you see in this particular stack trace to this discussion, but indeed I also spotted this issue when client disconnects abruptly. I used `wrk` to hit this problem. But many won't even see this problem if they use reverse-proxy with request/response buffering.

@Kludex IMHO, instead of closing comments on resolved issues, it would be nice to let community to leave pointers to the correct remedy.

Kudos to @kprzybyla for finding the workaround! 🙇 
It did work for me too with my `wrk` use.

However, there seems to be a large number of cases, when this error will be hitting the screen, because provided workarounds doesn't handle all situations. For example: if you hit `Ctrl+C` *before* running any HTTP calls that would leave orphan connections, the exception doesn't appear at all. But if you made at least 1 request with client that won't close a connection properly, and then you send `INT` or `TERM` signal, the app will hang in the exit state unless you send another `INT` signal again. And if you do send second `INT`, you'll get `asyncio.exceptions.CancelledError` error:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 366, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/usr/local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.9/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.9/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.9/site-packages/starlette/middleware/base.py", line 68, in __call__
    response = await self.dispatch_func(request, call_next)
  File "/app/./app/core/metrics.py", line 22, in dispatch
    return await call_next(request)
  File "/usr/local/lib/python3.9/site-packages/starlette/middleware/base.py", line 43, in call_next
    message = await recv_stream.receive()
  File "/usr/local/lib/python3.9/site-packages/anyio/streams/memory.py", line 102, in receive
    await receive_event.wait()
  File "/usr/local/lib/python3.9/asyncio/locks.py", line 226, in wait
    await fut
asyncio.exceptions.CancelledError
```

Just in case, my `requirements.txt`:
```
uvicorn==0.17.6
fastapi==0.78.0
```

Also maybe important to note, it doesn't matter whether you add a custom middleware or not. I tried to run `wrk` without my middleware (you can see it in the stacktrace as `/app/./app/core/metrics.py`). `asyncio.exceptions.CancelledError` is guaranteed to appear on `INT` signal if there are any connections hanging that client didn't manage to close properly.

I'd love to contribute to the resolution of the issue, but I'm afraid I'm quite useless here, except nagging to open an issue so others would be aware of the problem 😊 

NB: I would never look into Discussions if I wouldn't see @Kludex point out to do so in this comment: https://github.com/encode/starlette/issues/1255#issuecomment-1110675635

**Reply by Kludex** [reply]
> @Kludex IMHO, instead of closing comments on resolved issues, it would be nice to let community to leave pointers to the correct remedy.

We have numerous issues that come into light after years of inactivity, which are, in general, new issues - even if the issue is similar.

**Reply by bob-rove** [reply]
No worries, it was just a humble suggestion from not too much educated user of the tool 😇 
Great to see you guys are putting efforts into so much loved framework! Total respect for your work 👍

**Reply by Kludex** [reply]
> However, there seems to be a large number of cases, when this error will be hitting the screen, because provided workarounds doesn't handle all situations.

You added the `is_disconnected` from the suggestion, but you still have `RuntimeError` instead of `Disconnected`?

**Reply by bob-rove** [reply]
Adding `is_disconnected` does help to prevent `RuntimeError`, but `CancelledError` doesn't go away on `fastapi==0.78.0` using `starlette==0.19.1` when hitting `Ctrl+C` after `wrk`, unless I do a follow up `while true ; do curl ... ; done`, as explained in https://github.com/encode/starlette/discussions/1527#discussioncomment-2770368

Here is my middleware just in case:
```python
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response
from starlette.types import ASGIApp
from prometheus_client import Gauge, CONTENT_TYPE_LATEST, REGISTRY, generate_latest

METRICS_HEADERS = {'Content-Type': CONTENT_TYPE_LATEST}

REQUESTS_IN_PROGRESS = Gauge(
    'starlette_requests_in_progress',
    'Gauge of requests by method and path currently being processed',
)


class PrometheusMiddleware(BaseHTTPMiddleware):
    def __init__(self, app: ASGIApp) -> None:
        super().__init__(app)

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        REQUESTS_IN_PROGRESS.inc()
        try:
            return await call_next(request)
        except RuntimeError as e:
            if await request.is_disconnected():
                return Response(status_code=204)
            else:
                raise
        finally:
            REQUESTS_IN_PROGRESS.dec()


def metrics(request: Request) -> Response:
    return Response(generate_latest(REGISTRY), headers=METRICS_HEADERS)
```

and then I jus add it to the FastAPI stack with:
```python
app = FastAPI(title=PROJECT_NAME, debug=DEBUG, version=VERSION, openapi_url='/openapi.json')

from app.core.metrics import PrometheusMiddleware, metrics

app.add_middleware(PrometheusMiddleware)
app.add_route('/metrics', metrics)
```

**Reply by bob-rove** [reply]
Just to clarify on `request.is_disconnected()` - I tried exact patch from https://github.com/encode/starlette/discussions/1527#discussioncomment-2242438 and it is showing `starlette.requests.ClientDisconnect: Client disconnect while processing the request` after `wrk` has existed together with `asyncio.exceptions.CancelledError` on `Ctrl+C`. But no `RuntimeError`.

In other words - in my case `request.is_disconnected()` helps to get rid of any errors after `wrk` exists __only__ when it's put within custom middleware on `starlette==0.19.1` with `fastapi==0.78.0`. But even then `asyncio.exceptions.CancelledError` still persists on `Ctrl+C` while server is idle (ie: not under the load).

**Comment by Kludex** [comment]
To help here, you can try to see if the problem is solved using the `master` branch, as we've merged https://github.com/encode/starlette/pull/1609.

If it doesn't help, see if the problem is solved using https://github.com/encode/starlette/pull/1441.

**Reply by bob-rove** [reply]
Tried both suggestions - didn't help. This is what I observed in my use case (which seems to be pretty generic, except the use of `wrk` to hit the issue):

1. #1609 gives no visible change in behaviour (perhaps because I don't use streaming responses).
2. #1441 doesn't give `RuntimeError: No response returned.` error anymore after `wrk` exits, __but__ it seems to be adding a regression somewhere because:
    - first `Ctrl+C` shows the same behaviour as before (server hangs in the "exiting" state)
    - second `Ctrl+C` throws `asyncio.exceptions.CancelledError` error __and__ not stopping the server (I have to use `KILL` signal to terminate the process)

**Reply by bob-rove** [reply]
@Kludex Here is another interesting observation:
1. run `wrk -c 100 -d 1 ...`: after it exits, the server output is flooded with `RuntimeError: No response returned.`
2. run `while true ; do curl ... ; done`: for a while, the server output contains both - successful response code and error `RuntimeError: No response returned.`
3. hit `Ctrl+C` - server exits immediately with no trace of `CancelledError` exceptions

The errors amount is almost equal to the concurrency level I used (>95).
The stack trace of `RuntimeError`s on both stages is identical:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 366, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/usr/local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.9/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.9/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/usr/local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.9/site-packages/starlette/middleware/base.py", line 68, in __call__
    response = await self.dispatch_func(request, call_next)
  File "/app/./app/core/metrics.py", line 22, in dispatch
    return await call_next(request)
  File "/usr/local/lib/python3.9/site-packages/starlette/middleware/base.py", line 47, in call_next
    raise RuntimeError("No response returned.")
RuntimeError: No response returned.
```

So I suspect there might be not properly closed responses (couldn't find better words to explain a hypothesis) that are preventing the server from being shutdown successfully without draining them first or having `CancelledError` exception appearing (original behaviour).

If I apply #1441 and reproduce these 3 steps, then `curl` doesn't even produce any output - no requests are being processed (using `curl -m 1` shows `curl: (28) Operation timed out after 1002 milliseconds with 0 bytes received` error).

**Comment by Kludex** [comment]
I'm running the example provided above:
```python
import asyncio

from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
from starlette.routing import Route


async def homepage(request):
    await asyncio.sleep(5)
    return JSONResponse({"hello": "world"})


routes = [
    Route("/", homepage, name="homepage"),
]


class CustomHeaderMiddleware1(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        return await call_next(request)


class CustomHeaderMiddleware2(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        return await call_next(request)


all_middleware = [
    Middleware(CustomHeaderMiddleware1),
    Middleware(CustomHeaderMiddleware2),
]


app = Starlette(debug=True, routes=routes, middleware=all_middleware)
```
with: 
```bash
uvicorn main:app --reload --port 8001
```

And then I'm running:
```bash
wrk -c 100 -d 1 http://0.0.0.0:8001/
```

To be sure, I've added a counter, to see how many failures because of disconnection.

I've run this against Starlette from today, 30 days ago, and 120 days ago (master branch wise). It looks like there's no regression, the issue is the same. We never stop having this issue, it was always here.

I don't see anything on FastAPI that could have caused this issue to come into light more easily.

If someone has some more insights, please feel free to share.

**Reply by bob-rove** [reply]
I don't know to be honest if it is a regression or not, but using a custom middleware doesn't yield `RuntimeError` on this spec:
```
fastapi==0.68.2
starlette==0.14.2
```
but it does on this one:
```
fastapi==0.69.0
starlette==0.15.0
```

**Reply by Kludex** [reply]
That's the version we introduced `anyio`... But we fixed some stuff on the `BaseHTTPMiddleware` since that version. Also, FastAPI is already on 0.78.0...

It will be really appreciated if we can debug things here with pure Starlette. FastAPI pins Starlette, and that makes things hard here.

**Reply by bob-rove** [reply]
I understand - for me using `fastapi` is the easiest way to pin `starlette` version. Meaning - at least I understand what am I doing. Going other route will render me clueless as I'll have to dig into abstractions I have no knowledge about 😭 
I believe your assumption that FastAPI has no effect in the issue is pretty much correct.
Have you tried running pure `starlette` prior to `0.15.0` btw?

**Reply by robotadam** [reply]
I've been tracking this issue on my application as well (also fastapi with custom middleware). I ran the script above as @bob-rove suggested, using starlette `0.14.2`, and it does not fail. On `0.14.2` the server does not report any requests at all (which is a little strange — I'd expect a log line if the request was processed at all), and on  `master` it reliably triggers the `anyio.EndOfStream`->`RuntimeError` failure.

**Reply by Kludex** [reply]
> I've been tracking this issue on my application as well (also fastapi with custom middleware). I ran the script above as @bob-rove suggested, using starlette `0.14.2`, and it does not fail. On `0.14.2` the server does not report any requests at all (which is a little strange — I'd expect a log line if the request was processed at all), and on  `master` it reliably triggers the `anyio.EndOfStream`->`RuntimeError` failure.
> 

Yeah, at some point that exception was obfuscated.

**Comment by acjh** [comment]
This error is raised when the stream is closed due to [`task_group.cancel_scope.cancel()`](https://github.com/encode/starlette/blob/9a1e885cca104d7fe1b650956a1c8259e497600f/starlette/responses.py#L262) in `StreamingResponse.__call__.<locals>.wrap` after `listen_for_disconnect`, and cancellation check in [`await checkpoint()`](https://github.com/agronholm/anyio/blob/c1bd3b4ef897540193445cdbd929702a4e56ad6c/src/anyio/streams/memory.py#L207) of `MemoryObjectSendStream.send`.

I managed to make this error reproducible in 0.14.2 by partially emulating 0.15.0 logic: https://github.com/acjh/starlette/commit/37dd8ace69993bf437712bc54a2fdb93116ab918

I have submitted PR #1706 to shield send "http.response.start" from cancellation.

**Reply by robotadam** [reply]
I saw that you have the separate implementation in #1710, but that there was a discussion about potentially changing the method to an exception. Have you thought more about which way you might go? And is there anything you need from those of us watching the discussion? Extra testing, etc?

**Comment by kigawas** [comment, marked as answer]
This is fixed in https://github.com/encode/starlette/pull/1715
